torch/utils/data/_utils/pin_memory.py

```diff
@@ -24,26 +24,28 @@ def _pin_memory_loop(in_queue, out_queue, device_id, done_event, device):
     # See NOTE [ Data Loader Multiprocessing Shutdown Logic ] for details on the
     # logic of this function.
     while not done_event.is_set():
-        try:
-            r = in_queue.get(timeout=MP_STATUS_CHECK_INTERVAL)
-        except queue.Empty:
-            continue
-        idx, data = r
-        if not done_event.is_set() and not isinstance(data, ExceptionWrapper):
-            try:
-                data = pin_memory(data, device)
-            except Exception:
-                data = ExceptionWrapper(
-                    where="in pin memory thread for device {}".format(device_id))
-            r = (idx, data)
-        while not done_event.is_set():
+        # Make sure that we don't preserve any object from one iteration
+        # to the next
+        def one_step():
             try:
-                out_queue.put(r, timeout=MP_STATUS_CHECK_INTERVAL)
-                break
-            except queue.Full:
-                continue
-        del r  # save memory
```

Comment:
I feel like it would be simpler if we just do `r = None` at the end if we want to move the Tensor out of scope. WDYT?

But, the original problem that GIL is blocked still persist when `gc` tries to clean up the Tensor.

Reply:
> I feel like it would be simpler if we just do r = None at the end if we want to move the Tensor out of scope. WDYT?

The problem is that doing `r=None` here only clear the tuple itself. `data` is still a local variable that remain alive. So this line is doing pretty much nothing today.
We could do `del r, data` to solve this. But moving it into the function makes it more future-proof as all local state will be properly removed.

> But, the original problem that GIL is blocked still persist when gc tries to clean up the Tensor.

Not sure how this is linked?
If you're talking about expensive unmap that happen on these objects, https://github.com/pytorch/pytorch/pull/83623 should fix that.

Reply:
Yes I left it there on purpose but without any strong reason.
Re-creating it means that there is no captured variable kept alive by the function.

But that shouldn't happen here indeed. moving it out!